pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -90,22 +125,335 @@ func New(cfg Config,
+	level.Info(c.logger).Log("msg", "discovered tenants from bucket", "users", len(tenants))
+	return c.compactUsers(ctx, sc, tableName, tenants)
+}
+
+// See: https://github.com/grafana/mimir/blob/34852137c332d4050e53128481f4f6417daee91e/pkg/compactor/compactor.go#L566-L689
+func (c *Compactor) compactUsers(ctx context.Context, sc storeClient, tableName string, tenants []string) error {
+	// When starting multiple compactor replicas nearly at the same time, running in a cluster with
+	// a large number of tenants, we may end up in a situation where the 1st user is compacted by
+	// multiple replicas at the same time. Shuffling users helps reduce the likelihood this will happen.
+	rand.Shuffle(len(tenants), func(i, j int) {
+		tenants[i], tenants[j] = tenants[j], tenants[i]
+	})
+
+	// Keep track of tenants owned by this shard, so that we can delete the local files for all other users.
+	errs := multierror.New()
+	ownedTenants := make(map[string]struct{}, len(tenants))
+	for _, tenant := range tenants {
+		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+		if err := ctx.Err(); err != nil {
+			return fmt.Errorf("interrupting compaction of tenants: %w", err)
+		}
+
+		// Ensure the tenant ID belongs to our shard.
+		owned, err := c.sharding.OwnsTenant(tenant)
+		if err != nil {
+			c.metrics.compactionRunSkippedTenants.Inc()
+			level.Warn(c.logger).Log("msg", "unable to check if tenant is owned by this shard", "tenantID", tenant, "err", err)
+			continue
+		}
+		if !owned {
+			c.metrics.compactionRunSkippedTenants.Inc()
+			level.Debug(c.logger).Log("msg", "skipping tenant because it is not owned by this shard", "tenantID", tenant)
+			continue
+		}
+
+		ownedTenants[tenant] = struct{}{}
+
+		if err := c.compactTenantWithRetries(ctx, sc, tableName, tenant); err != nil {
+			switch {
+			case errors.Is(err, context.Canceled):
+				// We don't want to count shutdowns as failed compactions because we will pick up with the rest of the compaction after the restart.
+				level.Info(c.logger).Log("msg", "compaction for tenant was interrupted by a shutdown", "tenant", tenant)
+				return nil
+			default:
+				c.metrics.compactionRunFailedTenants.Inc()
+				level.Error(c.logger).Log("msg", "failed to compact tenant", "tenant", tenant, "err", err)
+				errs.Add(err)
+			}
+			continue
+		}
+
+		c.metrics.compactionRunSucceededTenants.Inc()
+		level.Info(c.logger).Log("msg", "successfully compacted tenant", "tenant", tenant)
+	}
+
+	return errs.Err()
+
+	// TODO: Delete local files for unowned tenants, if there are any.
+}
+
+func (c *Compactor) compactTenant(ctx context.Context, sc storeClient, tableName string, tenant string) error {
+	level.Info(c.logger).Log("msg", "starting compaction of tenant", "tenant", tenant)
+
+	// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	// TODO: Use ForEachConcurrent?
```

Comment:
Probably unnecessary to multithread here. There are better ways to introduce a configurable worker parallelism elsewhere (likely more centralized for reuse across tenants as well as periods and tsdbs). Happy to leave that for a followup PR if/when it proves warranted

Reply:
Nice, I removed the TODO. Note that we are already parallelizing at the table and tenant level.

Reply:
Parallelizing at the table level is probably ok for now. In the future, I'd rather have a worker pool for jobs (work is effectively a `(user, table, tsdb_shard_range)` tuple).

Reply:
Added a TODO for a followup PR.